danascheider/skyrim_inventory_management_frontend
# Select active game from dropdown

## Context

[**Display shopping lists scoped to games**](https://trello.com/c/a3EZXsPQ/108-display-shopping-lists-scoped-to-games)

In the [last PR](https://github.com/danascheider/skyrim_inventory_management_frontend/pull/99), I scoped the shopping list page to games so that, by default, shopping lists from the first (most recently updated) game are shown and other games can be selected by setting the `game_id` query param.

We want the user to be able to select a game through the UI and view the shopping lists for that game. This PR introduces a `GamesDropdown` component, which sets the query param when an option is selected.

## Changes

* `GamesDropdown` and `GamesDropdownOption` components
* New test for selecting a game from the games dropdown to view shopping lists for a specific game
* Incorporate the games dropdown into the `DashboardLayout` with an `includeGameSelect` prop

## Considerations

It was really important to me that the dropdown list be styleable, so I decided to make a custom component instead of using a native `select` tag. However, I also wanted the component to be accessible, so I invested the time to implement the right ARIA attributes, focus, and keyboard behaviour. There is a detailed list of requirements for the dropdown in the linked card and more information on the ARIA combobox role is available [here](https://www.w3.org/TR/wai-aria-practices-1.1/examples/combobox/aria1.1pattern/listbox-combo.html).

One piece of functionality I decided to leave out for now was the filtering. The logic turned out to be a little complex and I couldn't decide quite what behaviour I wanted. The biggest challenge was that, when the input was populated with the default game name, the list was automatically filtered to contain only that game - obviously bad UX. I considered matching only shorter strings and showing the full list in the dropdown if the string matched one game's name exactly, but that wasn't quite right either. At the end I left it the way it is in this PR - if you press enter or blur the component after typing the exact name of a game (case insensitive), it'll set the game to the one you typed. But there is no filtering of options at this time.

## Manual Test Cases

Testing this PR should centre on the dropdown component - the correct game should be selected and the query string set when you click on any of the options or press Enter or Space while an option is focussed. When you click outside the component or when focus moves to an element outside the component, the dropdown should be hidden. If you de-focus the component after typing the full and complete name of one of the games (case doesn't matter), that game should be selected in the query string and its shopping lists should be visible.

## Screenshots and GIFs

### Dropdown Collapsed

#### 1201px

<img width="1203" alt="Screen Shot 2021-07-24 at 4 15 02 pm" src="https://user-images.githubusercontent.com/5115928/126859862-6c0febb0-d755-41b5-8f32-a32411e1c04c.png">

#### 1025px

<img width="1026" alt="Screen Shot 2021-07-24 at 4 15 16 pm" src="https://user-images.githubusercontent.com/5115928/126859869-39d22558-ebb7-4d6d-afbb-25839549b824.png">

#### 769px

<img width="770" alt="Screen Shot 2021-07-24 at 4 15 34 pm" src="https://user-images.githubusercontent.com/5115928/126859872-4c7d6cac-ffaa-4161-b7df-f7b94d7257e8.png">

#### 601px

<img width="601" alt="Screen Shot 2021-07-24 at 4 15 49 pm" src="https://user-images.githubusercontent.com/5115928/126859875-0995c116-1607-4567-82d0-84aee6972def.png">

#### 425px

<img width="426" alt="Screen Shot 2021-07-24 at 4 16 03 pm" src="https://user-images.githubusercontent.com/5115928/126859880-2c63e7d7-dc4f-408a-94de-f20a3a8da86a.png">

### Dropdown Expanded

#### 1201px

<img width="1202" alt="Screen Shot 2021-07-24 at 4 16 31 pm" src="https://user-images.githubusercontent.com/5115928/126859890-415d63b7-8047-4bac-9a17-fc6857b2922b.png">

#### 1025px

<img width="1025" alt="Screen Shot 2021-07-24 at 4 16 45 pm" src="https://user-images.githubusercontent.com/5115928/126859897-d06d8ac8-72e2-4805-a9c5-fe5023d4237e.png">

#### 769px

<img width="770" alt="Screen Shot 2021-07-24 at 4 17 00 pm" src="https://user-images.githubusercontent.com/5115928/126859909-f493768b-a04a-460e-b46c-2c70e8d997cc.png">

#### 601px

<img width="601" alt="Screen Shot 2021-07-24 at 4 17 15 pm" src="https://user-images.githubusercontent.com/5115928/126859913-e49da67c-2c80-415d-832e-a695d6ae3be1.png">

#### 425px

<img width="425" alt="Screen Shot 2021-07-24 at 4 17 31 pm" src="https://user-images.githubusercontent.com/5115928/126859916-3c9cdc18-9c02-4743-ba20-f84bc2b07dd6.png">
